src/transformers/models/rag/modeling_rag.py

```diff
@@ -483,6 +484,7 @@ def __init__(
         question_encoder: Optional[PreTrainedModel] = None,
         generator: Optional[PreTrainedModel] = None,
         retriever: Optional = None,  # or maybe just use a `set_retriever(...)` method
+        ctx_encoder: Optional = None, #newly_added
```

Comment:
Thanks for extending RAG to support backpropagation to the ctx_encoder :)

I think it's ok to allow users to optionally provide the context encoder if they want to backpropagate the gradient to the context encoder, but I'd like your opinion on this @patrickvonplaten 

In this case maybe we can have a parameter somewhere to toggle the computation of the ctx_encoder embeddings using the ctx_encoder instead of taking the embeddings from the retriever that don't backprogagate to the ctx_encoder. What do you think ?

Reply:
@lhoestq  Thanks a lot for getting into this :). I was thinking the same.  This can be done by adding an if condition in the doc score calculation part. If there's no ctx_encoder calculate the doc scores with retrieved embedding.

Reply:
I think instead of optionally providing the `ctx_encoder` to `init` it would be nicer to make a setter method instead, something like:

```python
def set_context_encoder_for_training(...)
```

This is cleaner IMO. For composite models, I'd really like to limit the __init__ only to necessary components (being the `question_encoder`, `generator`, and `config`) and don't want to add any optional ones that are only used for training. It makes the model less understandable IMO.

What do you guys think?

Reply:
me too. Will do it.